crates/core/src/dht/actions/remove_queued_holding_workflow.rs

```diff
@@ -36,11 +57,36 @@ impl Future for RemoveQueuedHoldingWorkflowFuture {
             .register_waker(self.id.clone(), cx.waker().clone());
 
         if let Some(state) = self.context.try_state() {
-            if state.dht().has_exact_queued_holding_workflow(&self.pending) {
-                Poll::Pending
-            } else {
-                self.context.unregister_waker(self.id.clone());
-                Poll::Ready(())
+            let store = state.dht();
+            match self.state {
+                HoldingWorkflowQueueing::Processing => {
+                    if store.has_exact_queued_holding_workflow(&self.pending)
+                        || !store.has_exact_in_process_holding_workflow(&self.pending)
+                    {
+                        Poll::Pending
+                    } else {
+                        self.context.unregister_waker(self.id.clone());
+                        Poll::Ready(())
+                    }
+                }
+                HoldingWorkflowQueueing::Waiting(_) => {
+                    if !store.has_exact_queued_holding_workflow(&self.pending)
+                        || store.has_exact_in_process_holding_workflow(&self.pending)
+                    {
+                        Poll::Pending
+                    } else {
+                        self.context.unregister_waker(self.id.clone());
+                        Poll::Ready(())
+                    }
+                }
+                HoldingWorkflowQueueing::Done => {
+                    if store.has_exact_in_process_holding_workflow(&self.pending) {
```

Comment:
Can a PendingValidation ever go straight from Waiting (in the holding_queue) to Done? If not, this is fine, but if so, this isn't checking that.